File: 310.collections1/src/main/java/javax/MyList.java
package javax;

import java.util.List;
import java.util.ArrayList;

public class MyList {
    public static void main(String[] args) {
        assert(args.length >= 1);

        List<String> list = new ArrayList<String>();
        for(int i = 0; i < args.length; i++) {
            list.add(args[i]);
        }

        List<String> reversedTail = reverseTail(list);
        printList(reversedTail);
    }

    private static List<String> reverseTail(final List<String> in) {
        List<String> reversed = new ArrayList<String>();
        if (in.size() == 0) {
            return new ArrayList<String>();
        } else {
            reversed.add(0, in.get(0));
            for(int i = 1; i < in.size(); i++) {
                reversed.add(1, in.get(i));
            }
            return reversed;
        }
    }

    private static void printList(final List<String> in) {
        for(int i = 0; i < in.size(); i++) {
            if (i == 0) {
                System.out.print(in.get(i));
            } else {
                System.out.print(", " + in.get(i));
            }
        }
        System.out.println();
    }
}

File: 310.collections1/src/main/scala/solution1/MyList.scala
package solution1

import scala.collection.mutable.ListBuffer

object MyList extends App {
  assert(args.size >= 1, "Usage: MyList 1 2 3 ...")

  val list = args.toList
  val reversedTail = reverseTail(list)
  printList(reversedTail)

  private def reverseTail(in: List[String]): List[String] = {
    val reversed = new ListBuffer[String]()
    if (in.isEmpty) {
      reversed
    } else {
      for (i <- 1 until in.size) {
        in(i) +=: reversed
      }
      in(0) +=: reversed
    }
    reversed.toList
  }

  private def printList(in: List[String]): Unit = {
    for (i <- 0 until in.size) {
      if (i == 0) {
        print(in(i))
      } else {
        print(", " + in(i))
      }
    }
    println()
  }
}
